spec.bs

```diff
@@ -2137,6 +2139,35 @@ The output of running a Protected Audience `scoreAd()` script.
 
 </dl>
 
+This is returned via the following type:
+<pre class="idl">
+dictionary ScoreAdOutput {
+  required double desirability;
+  double bid;
+  boolean allowComponentAuction = false;
+};
+</pre>
+
+<div algorithm>
+
+To <dfn>process scoreAd output</dfn> given a |result|, a [=ECMAScript/Completion Record=]:
+  1. If |result| is an an [=abrupt completion=], return failure.
+  1. TODO: the code has special handling of where the return value is a straight up double, do we want to keep it? It is in the explainer.
+  1. Let |resultIDL| be the result of [=converted to an IDL value|converting=]
+          |result| to a {{ScoreAdOutput}}.
+  1. TODO: Shouldn't this be the completion's value that's being converted?
+  1. If an exception was caught in the previous step, return failure.
+  1. Let |scoreOutput| be a new [=score ad output=].

```

Comment:
After adding the ScoreAdOutput IDL, it seems we no longer need the [=score ad output=] struct.

For steps that used [=score ad output=] like https://github.com/WICG/turtledove/blob/main/spec.bs#L840, we can directly use the IDL value |resultIDL|, like "Let score be |resultIDL|[desirability]"

@domfarolino Is it correct?

Reply:
I sort of wanted it for currency stuff, since in my drafts I end up packaging the bid and bidCurrency fields into a single field in =score ad output=... though I guess it's not actually needed since they get moved out together anyway, but it does make the higher-level algorithm a bit simpler.

Reply:
I think that's correct yes. But feel free to take whichever path you prefer if you need the struct for other currency stuff as @morlovich says, or if removing will indeed simplify things (probably removing it makes sense).

Reply:
Hmm, there is also a need for currency check, but may be the right way for this is to have a (double, string) -> =bid with currency= algorithm, since that happens in other spots, too, so just calling a subroutine for that and looking for failure in both spots will be nice and brief.

Reply:
... How do I describe the fields if it goes away, though?

Reply:
I mean right now you have stuff like:
: <dfn>desirability</dfn>
:: A {{double}}. 
  Numeric score of the bid. Must be positive or the ad will be rejected. The winner of the auction
  is the bid which was given the highest score.
: <dfn>allow component auction</dfn>

... which describes individual fields; I am not sure what it would look like for the IDL.

Reply:
You can reference the IDL dictionary's double with `{{ScoreAdOutput/desirability}}`, does that answer your question? Sorry if I'm misunderstanding.

Reply:
I mean how do I do explanatory comments like this table in Fetch:
https://fetch.spec.whatwg.org/#dom-request-signal:~:text=s%20priority.-,request%20.%20method,-Returns%20request%E2%80%99s

... the infra [=score ad output=] type has quite a few explanatory comments which would be a shame to lose.

(And I am still kinda dithering on this whole thing since the conversion makes things slightly nicer...)

Reply:
OK, change made --- PTAL!

Reply:
(For some reason I am not seeing the last commit, though it's supposedly pushed, investigating...)

Reply:
Resolving this because https://github.com/WICG/turtledove/pull/598#discussion_r1222731649.